allennlp/interpret/influence_interpreters/influence_interpreter.py

```diff
@@ -0,0 +1,361 @@
+        The data loader used to load training instances.
+
+        !!! Note
+            This data loader is only used to call `DataLoader.iter_instances()`, so certain
+            `DataLoader` settings like `batch_size` will have no effect.
+
+    test_data_loader : `Lazy[DataLoader]`, optional (default = `Lazy(SimpleDataLoader)`)
+        The data loader used to load test instances when `interpret_from_file()` is called.
+
+        !!! Note
+            Like `train_data_loader`, this data loader is only used to call `DataLoader.iter_instances()`,
+            so certain `DataLoader` settings like `batch_size` will have no effect.
+
+    params_to_freeze : `Optional[List[str]]`, optional (default = `None`)
+        An optional list of strings, each of which should be a regular expression that matches
+        some parameter keys of the model. Any matching parameters will be have `requires_grad`
+        set to `False`.
+
+    cuda_device : `int`, optional (default = `-1`)
+        The index of GPU device we want to calculate scores on. If not provided, we uses `-1`
+        which correspond to using CPU.
+    """
+
+    default_implementation = "simple-influence"
+
+    def __init__(
+        self,
+        model: Model,
+        train_data_path: DatasetReaderInput,
+        train_dataset_reader: DatasetReader,
+        *,
+        test_dataset_reader: Optional[DatasetReader] = None,
+        train_data_loader: Lazy[DataLoader] = Lazy(SimpleDataLoader.from_dataset_reader),
+        test_data_loader: Lazy[DataLoader] = Lazy(SimpleDataLoader.from_dataset_reader),
+        params_to_freeze: Optional[List[str]] = None,
+        cuda_device: int = -1,
+    ) -> None:
+        self._model = model
+        self._vocab = model.vocab
+        self._test_dataset_reader = test_dataset_reader or train_dataset_reader
+        self._device = int_to_device(cuda_device)
+
+        self._train_loader = train_data_loader.construct(
+            reader=train_dataset_reader,
+            data_path=train_data_path,
+            batch_size=1,
+        )
+        self._train_loader.set_target_device(self._device)
+        self._train_loader.index_with(self._vocab)
+
+        self._lazy_test_data_loader = test_data_loader
+
+        self._model.to(self._device)
+        if params_to_freeze is not None:
+            for name, param in self._model.named_parameters():
+                if any([re.match(pattern, name) for pattern in params_to_freeze]):
+                    param.requires_grad = False
+
+        # This is not set until we actually run the calculation since some parameters might not be used.
+        self._used_params: List[torch.Tensor] = []
+        self._used_param_names: List[str] = []
+
+        # Load training instances, compute loss and gradients.
+        self.train_instances: List[TrainInstance] = []
+        self._model.train()
+        for instance in Tqdm.tqdm(
```

Comment:
Having all of this computation in a constructor makes me really nervous.  Maybe it's fine, I just try to avoid this design.  We handled this kind of initialization in hotflip by having [a method](https://github.com/allenai/allennlp/blob/0c7d60bcf008ef963a61ee6829420aec9ecf4372/allennlp/interpret/attackers/hotflip.py#L76-L83) that [gets called](https://github.com/allenai/allennlp/blob/0c7d60bcf008ef963a61ee6829420aec9ecf4372/allennlp/interpret/attackers/hotflip.py#L248-L249) the first time it's needed.